src/vos/lru_array.c

```diff
@@ -30,161 +30,350 @@
 #include "lru_array.h"
 
 static void
-evict_cb(struct lru_array *array, struct lru_entry *entry, uint32_t idx)
+evict_cb(struct lru_array *array, struct lru_sub *sub, struct lru_entry *entry,
+	 uint32_t idx)
 {
-	if (array->la_cbs.lru_on_evict == NULL)
+	uint32_t	real_idx;
+
+	if (array->la_cbs.lru_on_evict == NULL) {
+		/** By default, reset the entry */
+		memset(entry->le_payload, 0, array->la_record_size);
 		return;
+	}
+
+	real_idx = (sub->ls_array_idx << array->la_array_shift) + idx;
 
 	array->la_evicting++;
-	array->la_cbs.lru_on_evict(entry->le_payload, idx, array->la_arg);
+	array->la_cbs.lru_on_evict(entry->le_payload, real_idx, array->la_arg);
 	array->la_evicting--;
 }
 
 static void
-init_cb(struct lru_array *array, struct lru_entry *entry, uint32_t idx)
+init_cb(struct lru_array *array, struct lru_sub *sub, struct lru_entry *entry,
+	uint32_t idx)
 {
+	uint32_t	real_idx;
+
 	if (array->la_cbs.lru_on_init == NULL)
 		return;
 
-	array->la_cbs.lru_on_init(entry->le_payload, idx, array->la_arg);
+	real_idx = (sub->ls_array_idx << array->la_array_shift) + idx;
+
+	array->la_cbs.lru_on_init(entry->le_payload, real_idx, array->la_arg);
 }
 
 static void
-fini_cb(struct lru_array *array, struct lru_entry *entry, uint32_t idx)
+fini_cb(struct lru_array *array, struct lru_sub *sub, struct lru_entry *entry,
+	uint32_t idx)
 {
+	uint32_t	real_idx;
+
 	if (array->la_cbs.lru_on_fini == NULL)
 		return;
 
-	array->la_cbs.lru_on_fini(entry->le_payload, idx, array->la_arg);
+	real_idx = (sub->ls_array_idx << array->la_array_shift) + idx;
+
+	array->la_cbs.lru_on_fini(entry->le_payload, real_idx, array->la_arg);
+}
+
+static int
+array_alloc_one(struct lru_array *array, struct lru_sub *sub)
+{
+	struct lru_entry	*entry;
+	char			*payload;
+	size_t			 rec_size;
+	uint32_t		 nr_ents = array->la_idx_mask + 1;
+	uint32_t		 prev_idx = nr_ents - 1;
+	uint32_t		 idx;
+
+	rec_size = sizeof(*entry) + array->la_record_size;
+	D_ALLOC(sub->ls_table, rec_size * nr_ents);
+	if (sub->ls_table == NULL)
+		return -DER_NOMEM;
+
+	/** Add newly allocated ones to head of list */
+	d_list_del(&sub->ls_link);
+	d_list_add(&sub->ls_link, &array->la_free_sub);
+
+	payload = sub->ls_payload = &sub->ls_table[nr_ents];
```

Comment:
Because then I can't access it like an array.   We could do this but it would require a bunch of pointer math to look up the table entry rather than simple array indexing